### PR TITLE
Neuron model for causal lm partial support

### PR DESCRIPTION
**Pull request by dacorvo**
This is a very preliminary integration of a `NeuronModelForCausalLM` generic class to wrap neuronx optimized models.

For now, only `gpt2` model type is supported.

The following features are available:

- instantiate (export + compile) a neuron model from a `transformers` model_id  (`from_pretrained`),
- save the exported model locally (`save_pretrained`),
- instantiate (export + reload compilation artifacts) a neuron model from a local directory (`from_pretrained`),
- generate outputs (`generate`).

What is missing:
- between two generations, a specific `reset_generation` method must be called: I plan to remove that in a subsequent pull-request,
- we should check the compiler version when instantiating from a local path,
- the export should be available through the CLI,
- the new model class should have a corresponding `text-generation` pipeline.

What could be added if this is a use case:
- use cache (is it even possible since the compilation happens behind the scene in transformers-neuronx ?),
- support recompilation with different parameters when instantiating from a local path,

### CODE REVIEW

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:139]:
```suggestion
            model_path = model_dir.name
```

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:136]:
```suggestion
    def _get_neuron_paths(cls, model_dir: Union[str, Path, TemporaryDirectory]) -> Tuple[str, str, str]:
```

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:165]:
```suggestion
```

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:175]:
```suggestion
```

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:207]:
```suggestion
```

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:210]:
```suggestion
```

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:61]: It should inherint from `NeuronConfig` imo.

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:70]:
```suggestion
    def __init__(self, task: str):
```

Why allow for keyword arguments here?

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:72]:
```suggestion
            raise ModuleNotFoundError("The transformers-neuronx package is required. You can install it by running the following command: pip install transformers-neuronx")
```

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:80]:
```suggestion
        """
        Splits between kwargs that need to be passed when loading the transformers model
        and those that need to be passed to the neuron optimizer.
```

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:66]: Why are those values hardcoded here?

File: optimum/exporters/neuron/config.py
Comment (JingyaHuang) [inline optimum/exporters/neuron/config.py:61]: Would it make sense to inherit from `NeuronConfig` as I might have put / will put some type check of configs.

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:68]: Maybe we can refactor both like:

```python
NEURONX_CLASS`: str = "gpt2.model.GPT2ForSampling"
```

I do not think that adding the `transformers_neuronx` here is neeeded since it's implied and we can add it once when we load the actual class.

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [inline optimum/exporters/neuron/config.py:82]:
```suggestion
        model_kwargs = copy.deepcopy(kwargs)
```

File: optimum/neuron/modeling.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling.py:644]: property or attribute?
```suggestion
    def can_generate(self) -> bool:
```

File: optimum/neuron/modeling.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling.py:588]: nit: maybe add type annotation.

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:1]: Maybe we should move all the modeling classes and files for inference in a module.
Like `optimum.neuron.inference` or something.

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:130]: Not in favor of that because we want to keep the model config similar to what is used in Transformers.

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [inline optimum/neuron/modeling_decoder.py:49]: Why dont we inherint from `NeuronBaseModel`?

File: test_llm_optimum.py
Comment (michaelbenayoun) [inline test_llm_optimum.py:1]: What is this file?
Should it be removed?

**Review by michaelbenayoun — COMMENTED** [review]
Left some comments, but overall looks great!

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:61]: I considered this, but NeuronConfig is too specific to the models that were previously integrated: none of its methods nor class constants are actually required by the export code for causal LM models.
For the causl LM models, all we need is:
- some way to identify the corresponding transformers_neuronx model class,
- some way to identify the kwargs that need to be passed when calling from_pretrained on that class.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:66]: These are the default values for these parameters.

Note that for now every single model class in transformers_neuronx has a slightly different set of parameters, but these ones seem to be common. If this is confirmed that these parameters are really common to all models, then I could have declared them directly as named parameters in the _from_transformers method.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:68]: I agree that transformers_neuronx could be omitted and inserted just when the class is imported.

I cannot merge the two constants, but I will need to split them back at import, because using `importlib` I cannot import a single symbol without loading the module that contains it entirely. In other words, I cannot do that:
```
module = importlib.import_module(`transformers_neuronx.gpt2.model.GPT2ForSampling)
```
It will raise the following exception:
```
ModuleNotFoundError: No module named 'transformers_neuronx.gpt2.model.GPT2ForSampling'; 'transformers_neuronx.gpt2.model' is not a package
```
This is why I need first to load the module (I load it with importlib), then access the class as an attribute.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:70]: I think this is a left-over. I will remove it

File: optimum/neuron/modeling.py
Comment (dacorvo) [reply optimum/neuron/modeling.py:644]: This is actually a method of the PretrainedModel class that is called by the GenerationMixin.generate method.
Since we don't inherit from PretrainedModel we need to provide it explicitly.
This is a copy-paste of what is done in `onnxruntime` GenerationMixin.
I can definitely add the type annotation though.

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:49]: Again, this is because NeuronBaseModel is too specific: I looked carefully and there isn't a single method or attribute I could reuse here.

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:130]: I need to keep track of the task (to get the correct exporter), then I also need the named parameters that were used to configure the export.

I saw in NeuronBaseModel that the config is already extended with a lot of items, all named `neuron_XX`.
Do you mean you don't want to store anything neuron-specific in the config ? i certainly could serialize the neuron_config in its own file, like the generation_config, but since @JingyaHuang already had put a lot of stuff there I thought it was ok.

File: test_llm_optimum.py
Comment (dacorvo) [reply test_llm_optimum.py:1]: Yes, it is in its own "revertme" commit.

Comment (dacorvo) [reply test_llm_optimum.py:1]: I left it just for the time of the review because I don't know if it could be used as a base for a text-generation example (if any is required).

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [reply optimum/exporters/neuron/config.py:61]: Then I would move this to the `neuron/base.py` file since it can be considered as a base class in that sense.

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [reply optimum/exporters/neuron/config.py:66]: For me it could change, for instance `tp_degree` will need to be bigger for very big models and so on.

File: optimum/exporters/neuron/config.py
Comment (michaelbenayoun) [reply optimum/exporters/neuron/config.py:68]: Yes, I think it's okay to split them:

```python
module_name, attribute_name = self.NEURONX_CLASS.rsplit(".", maxsplit=1)
```

File: optimum/neuron/modeling_decoder.py
Comment (michaelbenayoun) [reply optimum/neuron/modeling_decoder.py:130]: I did not know. I think it's better to have everything in another file yes.
Do not want to extend the work on this but yes, in general we try not to change the original config.

File: optimum/neuron/modeling_decoder.py
Comment (JingyaHuang) [inline optimum/neuron/modeling_decoder.py:135]: What if I want to load from the hub, is it something feasible?

https://github.com/huggingface/optimum-neuron/blob/980dc5cfaa15b23ac3691b78f640a7182e3fd4df/optimum/neuron/modeling_base.py#L141-L143

File: optimum/neuron/modeling_decoder.py
Comment (JingyaHuang) [inline optimum/neuron/modeling_decoder.py:170]: For optimum modeling, we usually put the compilation under `from_transformers`, and `from_pretrained` will either load a pre-compiled model or compile it by calling `from_transformers`.

File: optimum/neuron/modeling_decoder.py
Comment (JingyaHuang) [inline optimum/neuron/modeling_decoder.py:200]: Why don't we serialize the compiled model anyway under `from_transformers` after compilation? Then here we have fewer cases to consider.

File: optimum/neuron/modeling.py
Comment (JingyaHuang) [inline optimum/neuron/modeling.py:599]: AFAIK, in optimum, we always `return_dict`, maybe just keep the convention.

**Review by JingyaHuang — COMMENTED** [review]
@dacorvo That's a very awesome Pr, thanks for the work. Congrats!

Left some nits, feel free to ping me for discussion or another review.

File: optimum/exporters/neuron/config.py
Comment (JingyaHuang) [reply optimum/exporters/neuron/config.py:61]: @dacorvo @michaelbenayoun if we finally decide to move this class to base.py, call it rather `NeuronDecoderConfig` .

File: optimum/exporters/neuron/config.py
Comment (JingyaHuang) [reply optimum/exporters/neuron/config.py:66]: If let's say gpt2 and llama not sharing the same fixed non-customizable parameters, may we put it under `GPT2NeuronConfig` directly?

File: optimum/neuron/modeling_decoder.py
Comment (JingyaHuang) [reply optimum/neuron/modeling_decoder.py:130]: I previously wanted to have a separate config for neuron-specific things, but to move fast I put them under pretrained config. @philschmid 

I now actually think it's not bad to have them in one place.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:61]: As explained it is not really possible, but i will move the class as requested.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:66]: Okay, this deserves a better explanation (I was very confused my self when i went through the _transformers-neuronx_ "export").

Firstable, one needs to understand that unlike other models, the export requires a preliminary "modeling/transform" step that happens before the tracing/compilation step.

In the current _transformers_neuronx_ implementation, this step must be performed every time from the source model, even when reloading a model that has already been compiled (this is why I need to copy the "checkpoint" model along the compiled model).

Unfortunately, this step requires the exact same parameters to be passed when the model is first compiled, then reloaded, and believe it or not, the information is not serialized. This is why I need to:
- list the parameters that need to be passed,
- serialize them myself in neuron_config,
- provide my own default values to avoid having one parameter set with a default value by the _transformers_neuronx_ code (note that I actually could have read back the values from the neuronx model `config` member: the one that is not serialized).

Comment (dacorvo) [reply optimum/exporters/neuron/config.py:66]: This is configurable.

Comment (dacorvo) [reply optimum/exporters/neuron/config.py:66]: > If let's say gpt2 and llama not sharing the same fixed non-customizable parameters, may we put it under `GPT2NeuronConfig` directly?

As explained in my previous answer, the parameters CAN be customized, but I need to provide my own default values. But yes, I could have put them under `GPT2NeuronConfig` directly.

Alternatively, as I already mentioned in another answer, I could at least declare batch_size, tp_degree and amp directly in the _from_transformers signature. These are really common to all models.

I thought n_positions was too, but I just double-checked and it is more used as a way to override the original config.n_positions.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:78]: Actually none of these parameters are mandatory, but if we don't provide values they will be set to default values.

As I said before, maybe I could make a difference between the three main parameters and the other ones.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:89]: Actually, they will never be any unused `kwargs`: the remaining `model_kwargs` will all be passed to `get_model_from_task`, and it will raise an error if one of them is not supported.
I think that the `modeling_base.py` export code ignores the model `kwargs` when calling `get_model_from_task`, but I cannot do that because some of them are really important like `low_cpu_memory`.

File: optimum/neuron/modeling.py
Comment (dacorvo) [reply optimum/neuron/modeling.py:533]: I think I did at line 572, but maybe I did it wrong ...

File: optimum/neuron/modeling.py
Comment (dacorvo) [reply optimum/neuron/modeling.py:599]: OK, so do I need to raise an error or add a warning and ignore if return_dict=False ?

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:127]: Here, this is not the compiled file yet, but just the original model (a.k.a checkpoint), that is serialized with each weight in a different file under a `pytorch.bin` directory (yes, the directory name is confuzing).

The compiled model (a.k.a compilation artifacts) is a pickle file containing all compiled binaries (one for each decoder layer). The file is named `neuron-program.pkl`.

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:130]: We need to decide for one solution or the other. I don't have any strong opinion myself, but if I put the neuron config in a separate file it will be just for these models for now.

Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:130]: Yes I can.

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:135]: I don't know, because my knowledge of the interactions with the hub is still very limited. This is something that would come in subsequent pull-requests (I might need guidance).

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:170]: This is what i did at first, but then I realized I need to apply the "modeling/transform" step also when loading a model from a local path. This is why it happens only later.

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:200]: Because in from_transformers we don't have the path to where it should be saved. But I agree that everybody will do a save_pretrained after a call to from_pretrained(export=True).

File: test_llm_optimum.py
Comment (dacorvo) [reply test_llm_optimum.py:1]: I added that in a commit with a revertme in the title: it will be removed later and was just here as an illustration of how the instantiation happens.

I now realize that you are more used to review files all at once than reviewing commits one by one, so this went under your radar and michael's.

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:61]: Done

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:66]: Done

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:68]: Done

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:72]: Done

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:78]: Done

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:82]: Done

File: optimum/exporters/neuron/config.py
Comment (dacorvo) [reply optimum/exporters/neuron/config.py:80]: Done

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py:81]: I added a comment.

File: optimum/neuron/modeling_decoder.py
Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py]: I'll stick with adding a neuron config inside the main config (@JingyaHuang did the same for othe rmodels).

Comment (dacorvo) [reply optimum/neuron/modeling_decoder.py]: Done

File: optimum/neuron/modeling.py
Comment (dacorvo) [reply optimum/neuron/modeling.py]: Done

**Review by michaelbenayoun — APPROVED** [review]
LGTM!
Thanks for the great work @dacorvo 

Can be merged once @JingyaHuang approved.

File: optimum/neuron/modeling_decoder.py
Comment (JingyaHuang) [reply optimum/neuron/modeling_decoder.py:146]: Never mind, we can get this in first and add the hub support in a separate PR. 
Basically it is just done through the `HfApi()` of `huggingface_hub`, as how it was done for other optimum modeling: https://github.com/huggingface/optimum-neuron/blob/980dc5cfaa15b23ac3691b78f640a7182e3fd4df/optimum/neuron/modeling_base.py#L141-L148

Feel free to ping me when you start to tackle this.

File: optimum/neuron/modeling.py
Comment (JingyaHuang) [reply optimum/neuron/modeling.py:608]: @dacorvo Personally don't feel it necessary, what about just remove `return_dict` from input arguments?

File: optimum/neuron/utils/version_utils.py
Comment (JingyaHuang) [inline optimum/neuron/utils/version_utils.py:51]: 👍

**Review by JingyaHuang — APPROVED** [review]
LGTM as well!

Amazing job, let's get it merged 🚀 !

**Comment by JingyaHuang** [comment]
Oh btw, let's add a guiding doc for decoder model as well (maybe in a separate PR): 
* https://huggingface.co/docs/optimum-neuron/guides/models
* https://github.com/huggingface/optimum-neuron/blob/main/docs/source/guides/models.mdx

(And also precise the limit like currently only works for gpt2 in the supported architecture: https://huggingface.co/docs/optimum-neuron/package_reference/configuration)

File: optimum/neuron/modeling.py
Comment (dacorvo) [reply optimum/neuron/modeling.py:608]: I tried to remove it, but I got an error and remembered why I had to put it there in the first place: this is because generate will pass it no matter what.

**Comment by dacorvo** [comment]
> Oh btw, let's add a guiding doc for decoder model as well (maybe in a separate PR):
> 
> * https://huggingface.co/docs/optimum-neuron/guides/models
> * https://github.com/huggingface/optimum-neuron/blob/main/docs/source/guides/models.mdx
> 
> (And also precise the limit like currently only works for gpt2 in the supported architecture: https://huggingface.co/docs/optimum-neuron/package_reference/configuration)

Ok, that will be my next pull-request, now that I finally figured out that I needed to actually edit the mdx files to get something generated 🤭.

File: optimum/neuron/modeling.py
Comment (JingyaHuang) [reply optimum/neuron/modeling.py:608]: Alright, I see, for onnxruntime it seems that someone added `**kwargs` to let it pass which I feel not the optimal solution either... but why not, let's either adopt that solution or set the default value to `True` at least?

Comment (dacorvo) [reply optimum/neuron/modeling.py:608]: Done

**Pull request merged**
